languages/bevy_mod_scripting_rune/src/lib.rs

```diff
@@ -0,0 +1,244 @@
+        // it compiles a file.
+        providers.attach_all(&mut context).unwrap();
+
+        let mut diagnostics = Diagnostics::new();
+
+        let mut sources = Sources::new();
+        sources
+            .insert(
+                Source::new(
+                    script_data.name,
+                    std::str::from_utf8(script).expect("Slice is not UTF-8"),
+                )
+                .map_err(|msg| ScriptError::FailedToLoad {
+                    script: script_data.name.into(),
+                    msg: msg.to_string(),
+                })?,
+            )
+            .map_err(|msg| ScriptError::FailedToLoad {
+                script: script_data.name.into(),
+                msg: msg.to_string(),
+            })?;
+
+        let result = rune::prepare(&mut sources)
+            .with_context(&context)
+            .with_diagnostics(&mut diagnostics)
+            .build();
+
+        if !diagnostics.is_empty() {
+            let mut writer = rune::termcolor::Buffer::no_color();
+
+            diagnostics
+                .emit(&mut writer, &sources)
+                .expect("Failed to write diagnostics to buffer");
+
+            return Err(ScriptError::SyntaxError {
+                script: script_data.name.into(),
+                msg: std::str::from_utf8(writer.as_slice())
+                    .expect("Slice was not UTF-8")
+                    .to_owned(),
+            });
+        }
+
+        let unit = result.expect("Failed to build Rune unit.");
+
+        let runtime_ctx = context
+            .runtime()
+            .expect("Failed to create Rune runtime context.");
+
+        Ok(RuneScriptContext {
+            unit: Arc::new(unit),
+            runtime_context: Arc::new(runtime_ctx),
+        })
+    }
+
+    fn setup_script(
+        &mut self,
+        script_data: &ScriptData,
+        ctx: &mut Self::ScriptContext,
+        providers: &mut APIProviders<Self>,
+    ) -> Result<(), ScriptError> {
+        providers.setup_all(script_data, ctx)
+    }
+
+    fn handle_events<'a>(
+        &mut self,
+        world: &mut World,
+        events: &[Self::ScriptEvent],
+        ctxs: impl Iterator<Item = (ScriptData<'a>, &'a mut Self::ScriptContext)>,
+        providers: &mut APIProviders<Self>,
+    ) {
+        // Safety:
+        // - we have &mut World access
+        // - we do not use the original reference again anywhere in this function
+        let world = unsafe { WorldPointerGuard::new(world) };
+
+        ctxs.for_each(|(script_data, ctx)| {
+            providers
+                .setup_runtime_all(world.clone(), &script_data, ctx)
+                .expect("Could not setup script runtime");
+
+            for event in events {
+                if !event.recipients().is_recipient(&script_data) {
+                    continue;
+                }
+
+                // TODO: should we store `Vm` in a script context?
+                let mut vm = Vm::new(ctx.runtime_context.clone(), ctx.unit.clone());
```

Comment:
I am not familiar with Rune, but this seems like what should be stored as the script context, copying the entire context, and creating a new stack is definitely not the way to do this :)

Reply:
Yeah, we should be storing a rune `Vm` per script as the script context. The only problem is that `Vm` doesn't implement `Send` or `Sync` which `ScriptHost::ScriptContext` requires. For whatever reason, sticking the vm in a `Mutex` doesn't work. 

Caching it could work. We could store `Vm`s in a non send resource or something. 

Do script contexts ever get accessed outside of the main thread or a single thread? From what I can tell, they don't. If they aren't, then I could wrap `Vm` in a new type and (unsafely) implement `Send` and `Sync` for it. So long as `handle_events` is the only system accessing them, this should be safe. 

(Also, the runtime context is inside an `Arc` so cloning it is cheap. Same with unit).

Reply:
Disregard my previous comment. Turns out that you can reuse a `Vm` for different units/contexts, which means we could store a single `Vm` in a non-send resource and reuse it for all script executions (I've never had to explore that corner of rune's API until now ¯\\\_(ツ)_/¯). I'm going to go ahead and implement it that way.